src/DotNet.Testcontainers/Containers/ResourceReaper.cs

```diff
@@ -0,0 +1,228 @@
+        .WithName($"testcontainers-ryuk-{sessionId:D}")
+        .WithImage(RyukImage)
+        .WithAutoRemove(true)
+        .WithCleanUp(false)
+        .WithExposedPort(RyukPort)
+        .WithPortBinding(0, RyukPort)
+        .WithBindMount("/var/run/docker.sock", "/var/run/docker.sock", AccessMode.ReadOnly)
+        .Build();
+
+      this.SessionId = sessionId;
+    }
+
+    /// <summary>
+    /// Gets the default <see cref="ResourceReaper" /> session id.
+    /// </summary>
+    /// <remarks>
+    /// The default <see cref="ResourceReaper" /> will start either on <see cref="GetAndStartDefaultAsync(System.Threading.CancellationToken)" />
+    /// or if a <see cref="ITestcontainersContainer" /> is configured with <see cref="ITestcontainersBuilder{TDockerContainer}.WithCleanUp" />.
+    /// </remarks>
+    public static Guid DefaultSessionId { get; }
+      = Guid.NewGuid();
+
+    /// <summary>
+    /// Gets the <see cref="ResourceReaper" /> session id.
+    /// </summary>
+    public Guid SessionId { get; }
+
+    /// <summary>
+    /// Starts and returns the default <see cref="ResourceReaper" /> instance.
+    /// </summary>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static async Task<ResourceReaper> GetAndStartDefaultAsync(CancellationToken ct = default)
+    {
+      // TODO: Use the State property?
+      if (defaultInstance != null && !defaultInstance.disposed)
+      {
+        return defaultInstance;
+      }
+
+      await DefaultLock.WaitAsync(ct)
+        .ConfigureAwait(false);
+
+      if (defaultInstance != null && !defaultInstance.disposed)
+      {
+        return defaultInstance;
+      }
+
+      try
+      {
+        defaultInstance = await GetAndStartNewAsync(DefaultSessionId, ct)
+          .ConfigureAwait(false);
+        return defaultInstance;
+      }
+      finally
+      {
+        DefaultLock.Release();
+      }
+    }
+
+    /// <summary>
+    /// Starts and returns a new <see cref="ResourceReaper" /> instance.
+    /// </summary>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static Task<ResourceReaper> GetAndStartNewAsync(CancellationToken ct = default)
+    {
+      return GetAndStartNewAsync(Guid.NewGuid(), ct);
+    }
+
+    /// <summary>
+    /// Starts and returns a new <see cref="ResourceReaper" /> instance.
+    /// </summary>
+    /// <param name="sessionId">The session id.</param>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static async Task<ResourceReaper> GetAndStartNewAsync(Guid sessionId, CancellationToken ct = default)
+    {
+      // TODO: Do we need to check if the session id is already in use?
```

Comment:
I would keep it simple and assume that this method is called only once with the same id. Also `resourceReaper.StartAsync(ct)` would throw if there is already a `ResourceReaper` container running with the same id as the container name is already in use in that case. We could add a test for this case to define this behavior as intended.

Reply:
I removed the todo.